src/hotspot/share/opto/constantTable.cpp

```diff
@@ -35,12 +35,15 @@
 bool ConstantTable::Constant::operator==(const Constant& other) {
   if (type()          != other.type()         )  return false;
   if (can_be_reused() != other.can_be_reused())  return false;
+  if (is_array() || other.is_array()) {
+    return is_array() && other.is_array() && _v._array == other._v._array;
+  }
   // For floating point values we compare the bit pattern.
   switch (type()) {
-  case T_INT:
-  case T_FLOAT:   return (_v._value.i == other._v._value.i);
-  case T_LONG:
-  case T_DOUBLE:  return (_v._value.j == other._v._value.j);
+  case T_INT:     return (_v._value.i == other._v._value.i);
+  case T_FLOAT:   return jint_cast(_v._value.f) == jint_cast(other._v._value.f);
```

Comment:
Why did you add `jint_cast`?

Does it mean the original impl for `T_FLOAD` wouldn't work as expected?

Reply:
They are the same. The current one uses type punning to do a bit cast. The updated one just makes the cast more explicit and consolidates the theoretical UB of type punning so that we can fix it more easily in the future.
Thanks.